src/composeappengine.cc

```diff
@@ -42,46 +43,67 @@ void ComposeAppEngine::remove(const App& app) {
 }
 
 bool ComposeAppEngine::isRunning(const App& app) const {
-  bool cmd_res{false};
-  std::string cmd_output;
-  std::tie(cmd_res, cmd_output) = cmd("cat " + (appRoot(app) / ComposeFile).string());
-  if (!cmd_res) {
-    LOG_WARNING << "Failed to parse App config: " << app.name;
-    return false;
-  }
+  try {
+    ComposeInfo compose((appRoot(app) / ComposeFile).string());
+    std::vector<Json::Value> services;
 
-  // calculate a number of container images that a given App consists of
-  std::string::size_type find_pos{0};
-  std::string::size_type line_pos{0};
-  int expected_container_number{0};
+    // get the expected services from the application
+    if (!compose.getServices(services)) {
+      LOG_WARNING << "App: " << app.name << ", no services in docker file!";
+      return false;
+    }
 
-  while ((find_pos = cmd_output.find("image:", line_pos)) != std::string::npos) {
-    if (cmd_output.substr(line_pos, (find_pos - line_pos)).find_first_of('#') == std::string::npos) {
-      ++expected_container_number;
+    // get all running services and their hashes for the given application
+    std::string req(
+        "ps -q --filter=status=running --filter=label=com.docker.compose.project=" + app.name +
+        R"( --format='{ "service": "{{.Label "com.docker.compose.service"}}", "hash": "{{.Label "io.compose-spec.config-hash"}}" },')");
+    std::string output;
+    bool ret{false};
+
+    std::tie(ret, output) = cmd(docker_ + req);
+    if (!ret || output.empty()) {
+      if (ret) {
+        LOG_INFO << "App: " << app.name << ", no services running";
+      } else {
+        LOG_WARNING << "App: " << app.name << ", failed to interrogate docker";
+      }
+      return false;
     }
-    line_pos = cmd_output.find('\n', find_pos);
-    ++find_pos;
-  }
 
-  // Get a number of running container images
-  std::tie(cmd_res, cmd_output) =
-      cmd(docker_ + "ps -q --filter=status=running --filter=label=com.docker.compose.project=" + app.name);
-  if (!cmd_res) {
-    LOG_WARNING << "Failed to get a list of App's containers: " << app.name;
-    return false;
-  }
+    // create a json array from the command output
+    std::string json("[" + output + "]");
```

Comment:
I wonder if these square brackets are really needed?

Reply:
yes, otherwise you will generate an exception